Extensions: implement drop-down menus

cwillisf opened the pull request:
### Resolves

Resolves #689 

### Proposed Changes

This change implements basic drop-down menu support for extension blocks. Dynamic menus -- ones which collect menu items fresh each time the menu opens -- are not yet supported.

Since I was in the area, I also fixed Boolean inputs and added angle picker support.

### Reason for Changes

This is another step toward implementing the whole extension spec.

Commit:
Implement drop-down menus for extension blocks

Also, add `ANGLE` argument type (like `NUMBER` but adds an angle picker)

Commit:
Extensions: fix Boolean inputs

Commit:
Fix `escapeHtml` error caused by rebase

ericrosenbaum reviewed (approved):
Looks good!

The menu field's color looks like it needs adjusting, but we can fix that later.

rschamp commented on src/engine/runtime.js:
```diff
@@ -400,19 +415,71 @@ class Runtime extends EventEmitter {
+        const menuId = this._makeExtensionMenuId(menuName, categoryInfo.id);
+
+        /** @TODO: support dynamic menus when 'menuItems' is a method name string (see extension spec) */
+        const options = menuItems.map(item => {
```
Seeing as this will be encountered frequently by developers, it might be nice to put some validation on the `menuItems` type and throw a warning if it's not an array/function that returns an array?

rschamp commented on src/engine/runtime.js:
```diff
@@ -400,19 +415,71 @@ class Runtime extends EventEmitter {
+                colour: categoryInfo.color1,
+                colourSecondary: categoryInfo.color2,
+                colourTertiary: categoryInfo.color3,
+                outputShape: ScratchBlocksConstants.OUTPUT_SHAPE_ROUND,
```
Should all menus be round, or are we going to support droppable vs non-droppable (square, don't accept blocks) dropdowns?

cwillisf replied:
Good question. My initial instinct is to make all menus accept blocks until someone complains :)
We could certainly add a flag controlling that. I'm not sure if that flag would make the most sense on the block or menu...

towerofnix replied:
Can I complain? :)

I mean, my honest opinion is that non-droppable menus should be done away with entirely, so that every menu is programmable - but I doubt that's going to happen (and there's reasons for it not to), so..

But practically I think it makes sense for extensions to be able to do what normal Scratch blocks can do. It seems like it should be possible for every block in the entire Scratch programming language to be converted to extensions (or at least "builtin extensions").

Commit:
Throw if extension tries to build a dynamic menu

Dynamic menus are not yet supported for extension menus, though they are
part of the extension spec. This change provides more thorough feedback
if an extension tries to register an unsupported dynamic menu.
